File: packages/entity-extraction/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/entityextraction/EntityExtractionHelper.java
package io.capawesome.capacitorjs.plugins.mlkit.entityextraction;

import androidx.annotation.NonNull;
import com.google.mlkit.nl.entityextraction.DateTimeEntity;
import com.google.mlkit.nl.entityextraction.Entity;
import com.google.mlkit.nl.entityextraction.EntityExtractorOptions;
import com.google.mlkit.nl.entityextraction.PaymentCardEntity;
import com.google.mlkit.nl.entityextraction.TrackingNumberEntity;
import java.util.Locale;

public class EntityExtractionHelper {

    private static final String ERROR_ENTITY_TYPE_INVALID = "entity type is invalid.";
    private static final String ERROR_LANGUAGE_INVALID = "language is invalid.";

    @NonNull
    public static String convertLanguageToModelIdentifier(@NonNull String language) throws Exception {
        switch (language) {
            case "ARABIC":
                return EntityExtractorOptions.ARABIC;
            case "CHINESE":
                return EntityExtractorOptions.CHINESE;
            case "DUTCH":
                return EntityExtractorOptions.DUTCH;
            case "ENGLISH":
                return EntityExtractorOptions.ENGLISH;
            case "FRENCH":
                return EntityExtractorOptions.FRENCH;
            case "GERMAN":
                return EntityExtractorOptions.GERMAN;
            case "ITALIAN":
                return EntityExtractorOptions.ITALIAN;
            case "JAPANESE":
                return EntityExtractorOptions.JAPANESE;
            case "KOREAN":
                return EntityExtractorOptions.KOREAN;
            case "POLISH":
                return EntityExtractorOptions.POLISH;
            case "PORTUGUESE":
                return EntityExtractorOptions.PORTUGUESE;
            case "RUSSIAN":
                return EntityExtractorOptions.RUSSIAN;
            case "SPANISH":
                return EntityExtractorOptions.SPANISH;
            case "THAI":
                return EntityExtractorOptions.THAI;
            case "TURKISH":
                return EntityExtractorOptions.TURKISH;
            default:
                throw new Exception(ERROR_LANGUAGE_INVALID);
        }
    }

    @NonNull
    public static String convertModelIdentifierToLanguage(@NonNull String modelIdentifier) {
        // The model identifiers (e.g. `EntityExtractorOptions.ENGLISH` = "english") are the
        // lowercase form of the corresponding `Language` enum values (e.g. "ENGLISH").
        return modelIdentifier.toUpperCase(Locale.ROOT);
    }

    public static int convertEntityTypeToFilterValue(@NonNull String entityType) throws Exception {
        switch (entityType) {
            case "ADDRESS":
                return Entity.TYPE_ADDRESS;
            case "DATE_TIME":
                return Entity.TYPE_DATE_TIME;
            case "EMAIL":
                return Entity.TYPE_EMAIL;
            case "FLIGHT_NUMBER":
                return Entity.TYPE_FLIGHT_NUMBER;
            case "IBAN":
                return Entity.TYPE_IBAN;
            case "ISBN":
                return Entity.TYPE_ISBN;
            case "MONEY":
                return Entity.TYPE_MONEY;
            case "PAYMENT_CARD":
                return Entity.TYPE_PAYMENT_CARD;
            case "PHONE":
                return Entity.TYPE_PHONE;
            case "TRACKING_NUMBER":
                return Entity.TYPE_TRACKING_NUMBER;
            case "URL":
                return Entity.TYPE_URL;
            default:
                throw new Exception(ERROR_ENTITY_TYPE_INVALID);
        }
    }

    @NonNull
    public static String convertEntityTypeToString(int type) {
        switch (type) {
            case Entity.TYPE_ADDRESS:
                return "ADDRESS";
            case Entity.TYPE_DATE_TIME:
                return "DATE_TIME";
            case Entity.TYPE_EMAIL:
                return "EMAIL";
            case Entity.TYPE_FLIGHT_NUMBER:
                return "FLIGHT_NUMBER";
            case Entity.TYPE_IBAN:
                return "IBAN";
            case Entity.TYPE_ISBN:
                return "ISBN";
            case Entity.TYPE_MONEY:
                return "MONEY";
            case Entity.TYPE_PAYMENT_CARD:
                return "PAYMENT_CARD";
            case Entity.TYPE_PHONE:
                return "PHONE";
            case Entity.TYPE_TRACKING_NUMBER:
                return "TRACKING_NUMBER";
            default:
                return "URL";
        }
    }

    @NonNull
    public static String convertDateTimeGranularityToString(int granularity) {
        switch (granularity) {
            case DateTimeEntity.GRANULARITY_YEAR:
                return "YEAR";
            case DateTimeEntity.GRANULARITY_MONTH:
                return "MONTH";
            case DateTimeEntity.GRANULARITY_WEEK:
                return "WEEK";
            case DateTimeEntity.GRANULARITY_DAY:
                return "DAY";
            case DateTimeEntity.GRANULARITY_HOUR:
                return "HOUR";
            case DateTimeEntity.GRANULARITY_MINUTE:
                return "MINUTE";
            case DateTimeEntity.GRANULARITY_SECOND:
                return "SECOND";
            default:
                return "UNKNOWN";
        }
    }

    @NonNull
    public static String convertPaymentCardNetworkToString(int network) {
        switch (network) {
            case PaymentCardEntity.CARD_AMEX:
                return "AMEX";
            case PaymentCardEntity.CARD_DINERS_CLUB:
                return "DINERS_CLUB";
            case PaymentCardEntity.CARD_DISCOVER:
                return "DISCOVER";
            case PaymentCardEntity.CARD_INTER_PAYMENT:
                return "INTER_PAYMENT";
            case PaymentCardEntity.CARD_JCB:
                return "JCB";
            case PaymentCardEntity.CARD_MAESTRO:
                return "MAESTRO";
            case PaymentCardEntity.CARD_MASTERCARD:
                return "MASTERCARD";
            case PaymentCardEntity.CARD_MIR:
                return "MIR";
            case PaymentCardEntity.CARD_TROY:
                return "TROY";
            case PaymentCardEntity.CARD_UNIONPAY:
                return "UNIONPAY";
            case PaymentCardEntity.CARD_VISA:
                return "VISA";
            default:
                return "UNKNOWN";
        }
    }

    @NonNull
    public static String convertParcelCarrierToString(int carrier) {
        switch (carrier) {
            case TrackingNumberEntity.CARRIER_FEDEX:
                return "FEDEX";
            case TrackingNumberEntity.CARRIER_UPS:
                return "UPS";
            case TrackingNumberEntity.CARRIER_DHL:
                return "DHL";
            case TrackingNumberEntity.CARRIER_USPS:
                return "USPS";
            case TrackingNumberEntity.CARRIER_ONTRAC:
                return "ONTRAC";
            case TrackingNumberEntity.CARRIER_LASERSHIP:
                return "LASERSHIP";
            case TrackingNumberEntity.CARRIER_ISRAEL_POST:
                return "ISRAEL_POST";
            case TrackingNumberEntity.CARRIER_SWISS_POST:
                return "SWISS_POST";
            case TrackingNumberEntity.CARRIER_MSC:
                return "MSC";
            case TrackingNumberEntity.CARRIER_AMAZON:
                return "AMAZON";
            case TrackingNumberEntity.CARRIER_I_PARCEL:
                return "I_PARCEL";
            default:
                return "UNKNOWN";
        }
    }
}

File: packages/entity-extraction/ios/Plugin/EntityExtractionHelper.swift
import Foundation
import MLKitEntityExtraction

// swiftlint:disable cyclomatic_complexity
@objc public class EntityExtractionHelper: NSObject {
    static let allModelIdentifiers: [EntityExtractionModelIdentifier] = [
        .arabic, .chinese, .dutch, .english, .french, .german, .italian, .japanese,
        .korean, .polish, .portuguese, .russian, .spanish, .thai, .turkish
    ]

    static func convertLanguageToModelIdentifier(_ language: String) throws -> EntityExtractionModelIdentifier {
        switch language {
        case "ARABIC":
            return .arabic
        case "CHINESE":
            return .chinese
        case "DUTCH":
            return .dutch
        case "ENGLISH":
            return .english
        case "FRENCH":
            return .french
        case "GERMAN":
            return .german
        case "ITALIAN":
            return .italian
        case "JAPANESE":
            return .japanese
        case "KOREAN":
            return .korean
        case "POLISH":
            return .polish
        case "PORTUGUESE":
            return .portuguese
        case "RUSSIAN":
            return .russian
        case "SPANISH":
            return .spanish
        case "THAI":
            return .thai
        case "TURKISH":
            return .turkish
        default:
            throw CustomError.languageInvalid
        }
    }

    static func convertModelIdentifierToLanguage(_ modelIdentifier: EntityExtractionModelIdentifier) -> String {
        switch modelIdentifier {
        case .arabic:
            return "ARABIC"
        case .chinese:
            return "CHINESE"
        case .dutch:
            return "DUTCH"
        case .french:
            return "FRENCH"
        case .german:
            return "GERMAN"
        case .italian:
            return "ITALIAN"
        case .japanese:
            return "JAPANESE"
        case .korean:
            return "KOREAN"
        case .polish:
            return "POLISH"
        case .portuguese:
            return "PORTUGUESE"
        case .russian:
            return "RUSSIAN"
        case .spanish:
            return "SPANISH"
        case .thai:
            return "THAI"
        case .turkish:
            return "TURKISH"
        default:
            return "ENGLISH"
        }
    }

    static func convertEntityTypeToFilter(_ entityType: String) throws -> EntityType {
        switch entityType {
        case "ADDRESS":
            return .address
        case "DATE_TIME":
            return .dateTime
        case "EMAIL":
            return .email
        case "FLIGHT_NUMBER":
            return .flightNumber
        case "IBAN":
            return .IBAN
        case "ISBN":
            return .ISBN
        case "MONEY":
            return .money
        case "PAYMENT_CARD":
            return .paymentCard
        case "PHONE":
            return .phone
        case "TRACKING_NUMBER":
            return .trackingNumber
        case "URL":
            return .URL
        default:
            throw CustomError.entityTypeInvalid
        }
    }

    static func convertEntityTypeToString(_ entityType: EntityType) -> String {
        switch entityType {
        case .address:
            return "ADDRESS"
        case .dateTime:
            return "DATE_TIME"
        case .email:
            return "EMAIL"
        case .flightNumber:
            return "FLIGHT_NUMBER"
        case .IBAN:
            return "IBAN"
        case .ISBN:
            return "ISBN"
        case .money:
            return "MONEY"
        case .paymentCard:
            return "PAYMENT_CARD"
        case .phone:
            return "PHONE"
        case .trackingNumber:
            return "TRACKING_NUMBER"
        default:
            return "URL"
        }
    }

    static func convertDateTimeGranularityToString(_ granularity: DateTimeGranularity) -> String {
        switch granularity {
        case .year:
            return "YEAR"
        case .month:
            return "MONTH"
        case .week:
            return "WEEK"
        case .day:
            return "DAY"
        case .hour:
            return "HOUR"
        case .minute:
            return "MINUTE"
        case .second:
            return "SECOND"
        default:
            return "UNKNOWN"
        }
    }

    static func convertPaymentCardNetworkToString(_ network: PaymentCardNetwork) -> String {
        switch network {
        case .amex:
            return "AMEX"
        case .dinersClub:
            return "DINERS_CLUB"
        case .discover:
            return "DISCOVER"
        case .interPayment:
            return "INTER_PAYMENT"
        case .JCB:
            return "JCB"
        case .maestro:
            return "MAESTRO"
        case .mastercard:
            return "MASTERCARD"
        case .mir:
            return "MIR"
        case .troy:
            return "TROY"
        case .unionpay:
            return "UNIONPAY"
        case .visa:
            return "VISA"
        default:
            return "UNKNOWN"
        }
    }

    static func convertParcelCarrierToString(_ carrier: ParcelTrackingCarrier) -> String {
        switch carrier {
        case .fedEx:
            return "FEDEX"
        case .UPS:
            return "UPS"
        case .DHL:
            return "DHL"
        case .USPS:
            return "USPS"
        case .ontrac:
            return "ONTRAC"
        case .lasership:
            return "LASERSHIP"
        case .israelPost:
            return "ISRAEL_POST"
        case .swissPost:
            return "SWISS_POST"
        case .MSC:
            return "MSC"
        case .amazon:
            return "AMAZON"
        case .iParcel:
            return "I_PARCEL"
        default:
            return "UNKNOWN"
        }
    }
}
